### PR TITLE
Implement Bid contract basics except inter-contract call interactions and testing infrastructure

### PR DESCRIPTION
**Pull request by CPerezz**
Adds the first implementation of the `Bid Contract` as well as the `RuskExternals` impl and tests for both things.

Issues left out of the scope of this PR:
- https://github.com/dusk-network/rusk/issues/164
- https://github.com/dusk-network/rusk/issues/163
- https://github.com/dusk-network/rusk/issues/162
- https://github.com/dusk-network/rusk/issues/161
- https://github.com/dusk-network/rusk/issues/160

### CODE REVIEW

**Comment by CPerezz** [comment]
After a deep investigation on why `extend_bid()` was failing. It's clear now that the root of the problem can, but it's not directly the `dusk-kelvin-map` (which we somehow knew since @julesdesmit was using it without problems).

The thing is that seems to happen something wrong on the encoding of the `schnorr::single_key::PublicKey` structure. 

If you see the following test:
```rust
#[test]
fn bid_contract_workflow_works() {
    // Init Env & Contract
    let store = MemStore::new();
    let wasm_contract = Wasm::new(Contract::new(), BYTECODE);
    let mut remote = Remote::new(wasm_contract, &store).unwrap();
    // Create CorrectnessCircuit Proof and send it
    let value = JubJubScalar::from(100000 as u64);
    let blinder = JubJubScalar::from(50000 as u64);
    let secret = JubJubAffine::from(GENERATOR_EXTENDED * value);
    let nonce = BlsScalar::one();
    let encrypted_data = PoseidonCipher::encrypt(
        &[value.into(), blinder.into()],
        &secret,
        &nonce,
    );
    let commitment = JubJubAffine::from(
        &(GENERATOR_EXTENDED * value) + &(GENERATOR_NUMS_EXTENDED * blinder),
    );
    let hashed_secret = sponge_hash(&[value.into()]);
    let secret_spend_key = SecretSpendKey::new(value, blinder);
    let psk = PublicSpendKey::from(&secret_spend_key);
    let stealth_addr = psk.gen_stealth_address(&value);
    let sk_r = secret_spend_key.sk_r(&stealth_addr);
    let proof = create_proof(commitment, value, blinder);
    let mut pub_inp_bytes = [0u8; 33];
    pub_inp_bytes[..].copy_from_slice(
        &PublicInput::AffinePoint(commitment, 0, 0).to_bytes(),
    );
    // Add leaf to the Contract's tree and get it's pos index back
    let mut cast = remote
        .cast_mut::<Wasm<Contract<MemStore>, MemStore>>()
        .unwrap();
    let (err, idx) = cast
        .transact(
            &Contract::bid(
                commitment,
                hashed_secret,
                nonce,
                encrypted_data,
                stealth_addr,
                PublicKey::from(stealth_addr.pk_r()),
                0u64,
                proof.clone(),
                proof.clone(),
                1,
                [PublicInput::AffinePoint(commitment, 0, 0).to_bytes()],
            ),
            store.clone(),
            RuskExternals { mem: None },
        )
        .unwrap();
    // If call succeeds, this should not fail.
    cast.commit().unwrap();
    assert!(err == false);
    assert!(idx == 0u64);

    // Sign the t_e (expiration) and call extend bid.
    let secret = SecretKey::from(sk_r);
    let signature =
        secret.sign(&mut rand::thread_rng(), BlsScalar::from(10u64));
    // Now that a Bid is inside the tree we should be able to extend it if the
    // correct signature is provided.
    let call_error = cast
        .transact(
            &Contract::<MemStore>::extend_bid(
                signature,
                PublicKey::from(stealth_addr.pk_r()),
            ),
            store.clone(),
            RuskExternals { mem: None },
        )
        .expect("Failed to call extend_bid method");

    // If call succeeds, this should not fail.
    cast.commit().expect("Commit couldn't be done");
    assert!(call_error == false);
```
    
We use `PublicKey::stealth_addr.pk_r()` as the public key that we will store in the map inside the Bid contract.
    
And inside the `bid()` fn I did the following test:
```rust
pub fn bid(
        &mut self,
        commitment: JubJubAffine,
        hashed_secret: BlsScalar,
        nonce: BlsScalar,
        encrypted_data: PoseidonCipher,
        stealth_address: StealthAddress,
        pk: PublicKey,
        // This will be avaliable inside of the contract scope.
        block_height: u64,
        correctness_proof: Proof,
        spending_proof: Proof,
        // TODO: Remove.
        pub_inputs_len: u8,
        pub_inputs: [[u8; 33]; 1],
    ) -> (bool, usize) {
        // Setup error flag to false
        let mut err_flag = false;
        // Verify proof of Correctness of the Bid.
        // TODO: Mask this unsafe somewhere else.
        unsafe {
            // TODO: We should avoid that.
            let proof_bytes = correctness_proof.to_bytes();
            match verify_proof(
                pub_inputs_len as usize,
                &pub_inputs[0][0],
                &proof_bytes[0],
                &crate::BID_CORRECTNESS_VK[0],
            ) {
                1i32 => (),
                0i32 => err_flag = true,
                // TODO: CHECK the panic! impl since it panics.
                _ => panic!("Parameter got malformed"),
            };
        };

        if err_flag {
            return (err_flag, usize::MAX);
        }

        // Compute maturity & expiration periods
        let expiration = block_height + MATURITY_PERIOD + EXPIRATION_PERIOD;
        let eligibility = block_height + MATURITY_PERIOD;
        // Generate the Bid instance
        let bid = Bid {
            encrypted_data,
            nonce,
            stealth_address,
            hashed_secret,
            c: commitment,
            eligibility,
            expiration,
            pos: u64::MAX,
        };
        // Panic and stop the execution if the same one-time-key tries to
        // bid more than one time.
        let idx = match self
            // TODO: Rename since it's confusing.
            .map()
            .get(PublicKey::from(bid.stealth_address.pk_r()))
        {
            // If no entries are found for this PK, add it to the map and the
            // tree
            Ok(None) => {
                // Append Bid to the tree and obtain the position of it.
                // TODO: Add an issue in Poseidon for the size obtention in the internal push impl.
                let idx = self.tree_mut().push(BidLeaf { bid });
                // Link the One-time PK to the idx in the Map
                // Since we checked on the `get` call that the value was not
                // inside, there's no need to check that this
                // returns `Ok(None)`. So we just unwrap
                // the `Result` and keep the `Option` untouched.
                self.map_mut()
                    .insert(PublicKey::from(bid.stealth_address.pk_r()), idx)
                    .unwrap();
                if pk != PublicKey::from(bid.stealth_address.pk_r()) {
                    err_flag = true;
                };
                idx
            }
            _ => {
                err_flag = true;
                // Return whatever
                usize::MAX
            }
        };

        // TODO: Inter-contract call
        (err_flag, idx)
    }
```

This returns an error which comes from the comparison:
>  if pk != PublicKey::from(bid.stealth_address.pk_r()) {
>                    err_flag = true;
>                };

Therefore, the two objects should be the same but aren't. And we need to see why.

File: contracts/bid/src/hosted.rs [hunk 1]
```diff
@@ -0,0 +1,110 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+//! Hosted interface for the Bid Contract.
+//!
+//! Here the interface of the contract that will run inside the hosted
+//! envoirnoment (WASM instance) is defined and implemented.
```
Comment (julesdesmit) [inline contracts/bid/src/hosted.rs:10]: environment*

**Review by julesdesmit — APPROVED** [review]
Just a typo, otherwise LGTM.

File: Cargo.toml [hunk 2]
```diff
@@ -66,7 +67,9 @@ canonical = "0.4"
...
 canonical_host = "0.4"
 
-[patch.crates-io]
-canonical = {git = "https://github.com/dusk-network/canonical", rev = "edf9efd"}
-canonical_derive = {git = "https://github.com/dusk-network/canonical", rev = "edf9efd"}
-canonical_host = {git = "https://github.com/dusk-network/canonical", rev = "edf9efd"}
\ No newline at end of file
+[dev-dependencies]
+bid_contract = {path = "contracts/bid"}
+bid-circuits = {path = "./contracts/bid/circuits"}
```
Comment (ZER0) [inline Cargo.toml:72]: A stupid thing, but be consistent. You either use always `./` or you don't.

File: contracts/bid/Cargo.toml [hunk 3]
```diff
@@ -1,9 +1,52 @@
...
+name = "bid_contract"
 version = "0.1.0"
 authors = ["CPerezz <c.perezbaro@gmail.com>"]
 edition = "2018"
 
-# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+[lib]
+crate-type = ["cdylib", "rlib"]
 
 [dependencies]
```
Comment (ZER0) [inline contracts/bid/Cargo.toml:10]: Skipping since it's not worthy to change any deps here.

File: contracts/bid/src/tree.rs [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+        PoseidonMaxAnnotation,
+        S,
+        u64,
+        BID_TREE_DEPTH,
+    > {
+        self.tree.iter_walk(block_height).unwrap()
+    }
+}
+
+// TODO: Fix block-height annotated search on the tree.
```
Comment (ZER0) [inline contracts/bid/src/tree.rs:107]: Can we remove the commented tests directly? I assume it won't work when we update the rest of the code anyway so I would prefer removing a commented test that is outdated. Unless there is a valid reason to keep around.

**Review by ZER0 — APPROVED** [review]
A couple of nits, maybe can be addressed before merging (especially the commented test)

File: contracts/bid/src/tree.rs (hunk 4 above)
Comment (CPerezz) [reply contracts/bid/src/tree.rs:107]: The thing is that we should not forget about implementing this test/functionallity. Will remove the test keeping the function and add an issue for it.

**Comment by CPerezz** [comment]
This PR fell outdated since the publish in `canonical` of a breaking change that wasn't respecting semver rules.

This broke the entire PR and therefore I need to re-write it.

**Pull request closed without merging**